XMonad/Util/ExtensibleState.hs

```diff
@@ -89,12 +91,20 @@ modifyStateExts f = State.modify $ \st -> st { extensibleState = f (extensibleSt
 -- | Apply a function to a stored value of the matching type or the initial value if there
 -- is none.
 modify :: (ExtensionClass a, XLike m) => (a -> a) -> m ()
-modify f = put . f =<< get
+modify = modifyM . (pure .)
 
--- | Like @modify@ but the result value is applied strictly in respect to
--- the monadic environment.
+-- | Apply an action to a stored value of the matching type or the initial value if there
+-- is none.
+modifyM :: (ExtensionClass a, XLike m) => (a -> m a) -> m ()
+modifyM f = put =<< f =<< get
+
+-- | Like @modify@ but the result value is forced to WHNF before being stored.
 modify' :: (ExtensionClass a, XLike m) => (a -> a) -> m ()
-modify' f = (put $!) . f =<< get
+modify' = modifyM' . (pure .)
```

Comment:
Trying to think really hard whether `(pure .)` introduces too much lazyness here, but probably not (at least the core looks very similar)

Reply:
Ah, right, now I remember.  The bottleneck was `put` and since we didn't even use a strict version of `=<<` before, this also should not matter now.  And yeah, there might be a rewrite rule for `=<< pure` anyways